### PR TITLE
Setup PETL with jobs "diagnosis," "services," and "visit"

### PR DESCRIPTION
**Pull request by MiguelAHPpih**

### CODE REVIEW

**Comment by mogoodrich** [comment]
@MiguelAHPpih  was this PR supposed to just have the PETL reports in it?  It seems to have a bunch of other stuff as well.

**Comment by MiguelAHPpih** [comment]
@mogoodrich It should be only about petl, but looks like it merged with the sample-collection branch. This was unexpected.

**Comment by mogoodrich** [comment]
Ah... is the sample-collection branch something we want to merge as well, or does the PR need to be reworked?

**Comment by brandones** [comment]
You can `git rebase -i origin/master` to just select the commits you want to include. And then `git push --force` this branch (since we will have rewritten its history).

**Comment by MiguelAHPpih** [comment]
I'm going to clear any issues it has and then merge it, if that's okay with you.
PD: Have you got any errors while loading the seslab.csv file again? Because in that PR you mentioned that it happened.

**Comment by brandones** [comment]
Sure, that's fine.

Me? I don't remember, can you link to the PR you're talking about?

**Comment by MiguelAHPpih** [comment]
I was refering to @mogoodrich , since he mentioned it in the sample-collection PR #26

**Comment by brandones** [comment]
Oh yeah, right, we already have a PR for that, and there are a bunch of JavaScript issues we were talking about there: https://github.com/PIH/openmrs-config-ses/pull/26 . It would be great if we could deal with that separately from all this PETL stuff.

**Comment by brandones** [comment]
@MiguelAHPpih If you just `git rebase` to get rid of the sample-collection commits, I think we can go ahead and merge this

**Comment by brandones** [comment]
@MiguelAHPpih Now this branch is deleting files, while master is editing those same files, causing conflicts.

Instead of deleting those files, you can `git checkout master [path/to/file]` to reset them to whatever's in master (make sure master is up to date). Or you can `git rebase -i master`.

File: configuration/pih/petl/jobs/general/visit/source.sql [hunk 1]
```diff
@@ -0,0 +1,53 @@
+select encounter_type_id into @tb_general from encounter_type where uuid= 'aa42cc6c-b9ee-4850-926c-dda4bb14d890';/*50 tb intake*/
```
Comment (brandones) [inline configuration/pih/petl/jobs/general/visit/source.sql:1]:
```suggestion
select encounter_type_id into @tb_general from encounter_type where uuid= 'aa42cc6c-b9ee-4850-926c-dda4bb14d890';  -- 50 tb intake
```

**Comment by MiguelAHPpih** [comment]
@brandones All done!. Thanks for the help

File: configuration/pih/petl/jobs/general/diagnosis/source.sql [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+frecuencia_mes int,
+frecuencia_manana int,
+frecuencia_tarde int
+);
+
+set @YEAR=YEAR(NOW());
+
+insert into temp_diagnosis (especialidad,frecuencia_dias,frecuencia_semana,frecuencia_mes,frecuencia_manana,frecuencia_tarde)
+select 'tuberculosis',(count(encounter_type))/251 as frecuencia_dia,
+(count(encounter_type)/(WEEK(MAX(encounter_datetime)) - WEEK(MIN(encounter_datetime))+1)) as frecuencia_mes,
```
Comment (brandones) [inline configuration/pih/petl/jobs/general/diagnosis/source.sql:20]:
```suggestion
(count(encounter_type)/(WEEK(MAX(encounter_datetime)) - WEEK(MIN(encounter_datetime))+1)) as frecuencia_semana,
```

File: configuration/pih/petl/jobs/general/diagnosis/source.sql [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+(count(encounter_type)/(MONTH(MAX(encounter_datetime))-MONTH(MIN(encounter_datetime))+1)) as frecuencia_mes,
+(select count(encounter_type)/251 from encounter where HOUR (encounter_datetime) BETWEEN 01 AND 11 AND MINUTE (encounter_datetime) BETWEEN 00 AND 59 )as frecuencia_manana,
+(select count(encounter_type)/251 from encounter where HOUR (encounter_datetime) BETWEEN 12 AND 23 AND MINUTE (encounter_datetime) BETWEEN 00 AND 59 ) as frecuencia_tarde
+from encounter
+where encounter_type=@tb_general and YEAR(encounter_datetime) = @YEAR;
+
+insert into temp_diagnosis (especialidad,frecuencia_dias,frecuencia_semana,frecuencia_mes,frecuencia_manana,frecuencia_tarde)
+select 'covid',(count(encounter_type))/251 as frecuencia_dia,
+(count(encounter_type)/(WEEK(MAX(encounter_datetime)) - WEEK(MIN(encounter_datetime))+1)) as frecuencia_mes,
```
Comment (brandones) [inline configuration/pih/petl/jobs/general/diagnosis/source.sql:29]:
```suggestion
(count(encounter_type)/(WEEK(MAX(encounter_datetime)) - WEEK(MIN(encounter_datetime))+1)) as frecuencia_semana,
```

File: configuration/pih/petl/jobs/general/diagnosis/source.sql [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+(count(encounter_type)/(MONTH(MAX(encounter_datetime))-MONTH(MIN(encounter_datetime))+1)) as frecuencia_mes,
+(select count(encounter_type)/251 from encounter where HOUR (encounter_datetime) BETWEEN 01 AND 11 AND MINUTE (encounter_datetime) BETWEEN 00 AND 59 )as frecuencia_manana,
+(select count(encounter_type)/251 from encounter where HOUR (encounter_datetime) BETWEEN 12 AND 23 AND MINUTE (encounter_datetime) BETWEEN 00 AND 59 ) as frecuencia_tarde
+from encounter
+where encounter_type=@covid_general and YEAR(encounter_datetime) = @YEAR;
+
+insert into temp_diagnosis (especialidad,frecuencia_dias,frecuencia_semana,frecuencia_mes,frecuencia_manana,frecuencia_tarde)
+select 'hiv',(count(encounter_type))/251 as frecuencia_dia,
+(count(encounter_type)/(WEEK(MAX(encounter_datetime)) - WEEK(MIN(encounter_datetime))+1)) as frecuencia_mes,
```
Comment (brandones) [inline configuration/pih/petl/jobs/general/diagnosis/source.sql:38]:
```suggestion
(count(encounter_type)/(WEEK(MAX(encounter_datetime)) - WEEK(MIN(encounter_datetime))+1)) as frecuencia_semana,
```

File: configuration/pih/petl/jobs/general/diagnosis/source.sql [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+);
+
+set @YEAR=YEAR(NOW());
+
+insert into temp_diagnosis (especialidad,frecuencia_dias,frecuencia_semana,frecuencia_mes,frecuencia_manana,frecuencia_tarde)
+select 'tuberculosis',(count(encounter_type))/251 as frecuencia_dia,
+(count(encounter_type)/(WEEK(MAX(encounter_datetime)) - WEEK(MIN(encounter_datetime))+1)) as frecuencia_mes,
+(count(encounter_type)/(MONTH(MAX(encounter_datetime))-MONTH(MIN(encounter_datetime))+1)) as frecuencia_mes,
+(select count(encounter_type)/251 from encounter where HOUR (encounter_datetime) BETWEEN 01 AND 11 AND MINUTE (encounter_datetime) BETWEEN 00 AND 59 )as frecuencia_manana,
+(select count(encounter_type)/251 from encounter where HOUR (encounter_datetime) BETWEEN 12 AND 23 AND MINUTE (encounter_datetime) BETWEEN 00 AND 59 ) as frecuencia_tarde
```
Comment (brandones) [inline configuration/pih/petl/jobs/general/diagnosis/source.sql:23]: Tal vez queriras excluyir 00:00 y 23:59, que son horas associadas con encuentros sin hora exacto

File: configuration/pih/petl/jobs/general/diagnosis/source.sql [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+frecuencia_tarde int
+);
+
+set @YEAR=YEAR(NOW());
+
+insert into temp_diagnosis (especialidad,frecuencia_dias,frecuencia_semana,frecuencia_mes,frecuencia_manana,frecuencia_tarde)
+select 'tuberculosis',(count(encounter_type))/251 as frecuencia_dia,
+(count(encounter_type)/(WEEK(MAX(encounter_datetime)) - WEEK(MIN(encounter_datetime))+1)) as frecuencia_mes,
+(count(encounter_type)/(MONTH(MAX(encounter_datetime))-MONTH(MIN(encounter_datetime))+1)) as frecuencia_mes,
+(select count(encounter_type)/251 from encounter where HOUR (encounter_datetime) BETWEEN 01 AND 11 AND MINUTE (encounter_datetime) BETWEEN 00 AND 59 )as frecuencia_manana,
```
Comment (brandones) [inline configuration/pih/petl/jobs/general/diagnosis/source.sql:22]:
```suggestion
(select count(encounter_type)/251 from encounter where HOUR (encounter_datetime) BETWEEN 00 AND 11 AND MINUTE (encounter_datetime) BETWEEN 00 AND 59 )as frecuencia_manana,
```

no? Para la hora despues de medianoche?

File: configuration/pih/petl/jobs/general/services/source.sql [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+frecuencia_manana int,
+frecuencia_tarde int
+);
+
+
+set @YEAR=YEAR(NOW());
+
+insert into temp_services (especialidad,frecuencia_dias,frecuencia_semana,frecuencia_mes,frecuencia_manana,frecuencia_tarde)
+select 'examenes',(count(encounter_type))/251 as frecuencia_dia,
+(count(encounter_type)/(WEEK(MAX(encounter_datetime)) - WEEK(MIN(encounter_datetime))+1)) as frecuencia_mes,
```
Comment (brandones) [inline configuration/pih/petl/jobs/general/services/source.sql:19]:
```suggestion
(count(encounter_type)/(WEEK(MAX(encounter_datetime)) - WEEK(MIN(encounter_datetime))+1)) as frecuencia_semana,
```

File: configuration/pih/petl/jobs/general/diagnosis/source.sql [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+(select count(encounter_type)/251 from encounter where HOUR (encounter_datetime) BETWEEN 12 AND 23 AND MINUTE (encounter_datetime) BETWEEN 00 AND 59 ) as frecuencia_tarde
+from encounter
+where encounter_type=@tb_general and YEAR(encounter_datetime) = @YEAR;
```
Comment (brandones) [inline configuration/pih/petl/jobs/general/diagnosis/source.sql:25]: El calculacion que quieres es para este año del calendario?

En este caso, debes adaptar el numero de dias de trabajo (251) para ser el numero aproximo de estos dias que han pasado, no? Algo como `DATEDIFF(NOW(), YEAR(NOW) + "-1-1") / 365 * 251`?

O quieres calcular por los ultimos 365 dias?

**Comment by brandones** [comment]
@MiguelAHPpih Hay unos errores y unas preguntas.

File: configuration/pih/petl/jobs/general/diagnosis/source.sql (hunk 8 above)
Comment (MiguelAHPpih) [reply configuration/pih/petl/jobs/general/diagnosis/source.sql:25]: Que hace el "-1-1"(puedes explicarlo en ingles para mayor precisión)?
PD: Gracias por el tip, la idea era al principio esa pero no encontré un modo de aplicarla en el momento..

Comment (brandones) [reply configuration/pih/petl/jobs/general/diagnosis/source.sql:25]: `YEAR(NOW) + "-1-1"` debe producir `2021-1-1`, la fecha.

File: configuration/pih/petl/jobs/general/diagnosis/source.sql (hunk 5 above)
Comment (MiguelAHPpih) [reply configuration/pih/petl/jobs/general/diagnosis/source.sql:23]: Incluyo las 24 horas para contar todos los encuentros que se hayan hecho, ya que por motivos de red algunas citas puede que no se registren ese mismo día en que se realizo el registro y tenga que hacerlo el siguiente.
No se si eso te parece correcto.

File: configuration/pih/petl/jobs/general/diagnosis/source.sql (hunk 8 above)
Comment (MiguelAHPpih) [reply configuration/pih/petl/jobs/general/diagnosis/source.sql:25]: ¿Se supone que arroja un null en lugar de un decimal o un integer ?. Además, ejecutar `YEAR (NOW) +" -1-1 "`, muestra solo el año 2020 en lugar del formato de fecha completo y el año correcto. Quizás sería mejor dejarlo con "2021-01-01" por ahora.¿?

Comment (brandones) [reply configuration/pih/petl/jobs/general/diagnosis/source.sql:25]: Oh noooo, SQL es un fracaso. Hace `CONCAT(YEAR(NOW()), '-01-01')`

Comment (MiguelAHPpih) [reply configuration/pih/petl/jobs/general/diagnosis/source.sql:26]: All done!

File: configuration/pih/petl/jobs/general/diagnosis/source.sql [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+(select count(encounter_type)/@diaslaborales from encounter where HOUR (encounter_datetime) BETWEEN 01 AND 11 AND MINUTE (encounter_datetime) BETWEEN 00 AND 59 )as frecuencia_manana,
+(select count(encounter_type)/@diaslaborales from encounter where HOUR (encounter_datetime) BETWEEN 12 AND 23 AND MINUTE (encounter_datetime) BETWEEN 00 AND 59 ) as frecuencia_tarde
+from encounter
+where encounter_type=@tb_general and YEAR(encounter_datetime) = @YEAR;
+
+insert into temp_diagnosis (especialidad,frecuencia_dias,frecuencia_semana,frecuencia_mes,frecuencia_manana,frecuencia_tarde)
+select 'covid',(count(encounter_type))/@diaslaborales as frecuencia_dia,
+(count(encounter_type)/(WEEK(MAX(encounter_datetime)) - WEEK(MIN(encounter_datetime))+1)) as frecuencia_semana,
+(count(encounter_type)/(MONTH(MAX(encounter_datetime))-MONTH(MIN(encounter_datetime))+1)) as frecuencia_mes,
+(select count(encounter_type)/@diaslaborales from encounter where HOUR (encounter_datetime) BETWEEN 01 AND 11 AND MINUTE (encounter_datetime) BETWEEN 00 AND 59 )as frecuencia_manana,
```
Comment (brandones) [inline configuration/pih/petl/jobs/general/diagnosis/source.sql:32]: We're still missing the hour after midnight, from 0:00 to 0:59. But you might want to just do 0:01 to 0:59 to exclude dates with no actual time associated with them.

File: configuration/pih/petl/jobs/general/diagnosis/source.sql [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+);
+
+set @YEAR=YEAR(NOW());
+set @diaslaborales=datediff(NOW(),CONCAT(YEAR(NOW()), '-01-01'))/365*251;
+
+insert into temp_diagnosis (especialidad,frecuencia_dias,frecuencia_semana,frecuencia_mes,frecuencia_manana,frecuencia_tarde)
+select 'tuberculosis',(count(encounter_type))/@diaslaborales as frecuencia_dia,
+(count(encounter_type)/(WEEK(MAX(encounter_datetime)) - WEEK(MIN(encounter_datetime))+1)) as frecuencia_semana,
+(count(encounter_type)/(MONTH(MAX(encounter_datetime))-MONTH(MIN(encounter_datetime))+1)) as frecuencia_mes,
+(select count(encounter_type)/@diaslaborales from encounter where HOUR (encounter_datetime) BETWEEN 01 AND 11 AND MINUTE (encounter_datetime) BETWEEN 01 AND 59 )as frecuencia_manana,
```
Comment (brandones) [inline configuration/pih/petl/jobs/general/diagnosis/source.sql:23]: I think you want

```suggestion
(select count(encounter_type)/@diaslaborales from encounter where HOUR (encounter_datetime) BETWEEN 00 AND 11 AND MINUTE (encounter_datetime) BETWEEN 01 AND 59 )as frecuencia_manana,
```

or

```suggestion
(select count(encounter_type)/@diaslaborales from encounter where HOUR (encounter_datetime) BETWEEN 00 AND 11 AND MINUTE (encounter_datetime) BETWEEN 00 AND 59 )as frecuencia_manana,
```

depending on whether or not you want to count encounters that were recorded retrospectively.

**Review by brandones — APPROVED** [review]
Looks great @MiguelAHPpih , thanks!

**Pull request merged**